### PR TITLE
Update Whoami from Python 2.7 to Python 3.7

### PR DESCRIPTION
**Pull request by nulven**
Updated requirements.txt versions
Changed filter and map objects to lists
Fixed print statements
Updated deprecated libraries

### CODE REVIEW

File: whoami/defextract.py [hunk 1]
```diff
@@ -211,8 +211,8 @@ def extract_assertion(raw_sentence, subject=[]):
 def extract_which_np(root):
     global ners
     try:
-        subject_head = filter(sdep.is_subject, root.children)[0]
-        determiner = filter(sdep.is_determiner, subject_head.children)[0]
+        subject_head = list(filter(sdep.is_subject, root.children))[0]
+        determiner = list(filter(sdep.is_determiner, subject_head.children))[0]
```
Comment (pondpremt) [inline whoami/defextract.py]: How about `next(filter(..., ...))`?

https://stackoverflow.com/questions/2361426/get-the-first-item-from-an-iterable-that-matches-a-condition

File: whoami/defextract.py [hunk 2]
```diff
@@ -237,8 +237,8 @@ def extract_what_is_np(root):
         return None
 
     try:
-        question_word = filter(wtype.is_wh_noun, np_dependents)[0]
-        definition = filter(lambda t: t != question_word, np_dependents)[0]
+        question_word = list(filter(wtype.is_wh_noun, np_dependents))[0]
+        definition = [t for t in np_dependents if t != question_word][0]
```
Comment (pondpremt) [inline whoami/defextract.py]: `next(t for t in np_dependents if t != question_word)`

File: whoami/postprocess.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ def merge_hyphen(node):
         if rel.is_hyphen(child):
             nodes_merged.append(child)
             nodes_removed.append(child)
-    map(node.remove_modifier, nodes_removed)
+    list(map(node.remove_modifier, nodes_removed))
```
Comment (pondpremt) [inline whoami/postprocess.py]: I think a plan old `for` loop might show the intent better

File: whoami/postprocess.py [hunk 4]
```diff
@@ -243,15 +242,15 @@ def get_common_hypernym(nodes):
...
+    for i in range(len(nodes)):
+        for j in range(i + 1, len(nodes)):
             common_hypernym = nodes[i].common_hypernym(nodes[j])
             if common_hypernym is not None:
                 if common_hypernym not in hypernyms:
                     hypernyms[common_hypernym] = 0
                 hypernyms[common_hypernym] += 1
     if hypernyms:
-        return max(hypernyms.iterkeys(), key=(lambda key: hypernyms[key]))
+        return max(iter(list(hypernyms.keys())), key=(lambda key: hypernyms[key]))
```
Comment (pondpremt) [inline whoami/postprocess.py]: My Python is a little rusty but shouldn't `max(hypernyms.keys(), ...)` do the job?

File: whoami/preprocess.py [hunk 5]
```diff
@@ -61,7 +61,7 @@ def replace_subject(sentence, subject=[]):
     the lookup table that maps the substitution words to the original strings.
     """
     substitutions = {}
-    synonyms = map(unicode.lower, filter(bool, subject))
+    synonyms = list(map(str.lower, list(filter(bool, subject))))
```
Comment (pondpremt) [inline whoami/preprocess.py]: I don't think the `list` calls are necessary; `synonyms` can be left as an iterable. So:

```
 synonyms = map(str.lower, filter(bool, subject))
```

File: whoami/defextract.py [hunk 6]
```diff
@@ -155,10 +155,10 @@ def extract_assertion(raw_sentence, subject=[]):
     ParseError if the sentence cannot be parsed as an is-a definition sentence.
     """
     global ners
-
+
```
Comment (pondpremt) [inline whoami/defextract.py]: I recommend setting your text editor to remove trailing spaces on save :)

File: whoami/preprocess.py (hunk 5 above)
Comment (nulven) [reply whoami/preprocess.py]: I'm not entirely sure of the reason why but as an iterable replace(sentence) returns incomplete results. I believe the iterator stops the for loop without going through duplicates.

**Pull request merged**